### ISSUE TITLE
April 5 - 9 - Build and Deploy applications to Azure by using GitHub Actions

### ISSUE THREAD

**Issue by mebaumb** [microsoft/ReactorPlanning#231]
# Event Information
| Category | Details |
|-----------|---------|
| Reactor Topic |  DevOps and Dev Tools |
| Monthly Topic (if applicable) | <Topic taken from [Azure Dev Marketing](https://aka.ms/DevEdCalFY21H1)> |
| Weekly Topic (if applicable) |  |
| Workshop Topic |  |
| Workshop Duration | | 
| Title | Build & Deploy Applications to Azure by Using GitHub Actions |
| Description | GitHub Actions are powerful workflow automations that support your projects success. You can use these automations in every step of the development project, from delivery to deployment. In this workshop we will learn how to use GitHub Actions to ensure continuous delivery (CD) of an application. We'll start by exploring options for triggering a CD GitHub Workflow, then control how the workflow is executed, and deploy the application to Microsoft Azure using a GitHub Deploy action. Afterwards, we'll do some data management while exploring GitHub Secrets and some additional GitHub Actions. To be successful in this workshop, you should have an understanding of GitHub Actions and continuous integration. |

# Content
https://aka.ms/GithubActionsCD-8

# Workshop(s) details
| Date |  Local Start Time |  Location |  Speaker | SharePoint ID | Platform |
|------|-------------------|-----------|----------|---------------|----------|
| April 6, 2021 | 6:30 PM | Bangalore | Devendra Rawat | 13040 |----------|
| April 6, 2021 | 4:00 PM | US/CA/AU | Allen Sanders | 13041 |----------|
| April 7, 2021 | 7:00 PM | Sao Paulo | Flavio Figueiredo | 13042 |----------|
| April 7, 2021 | 7:30 PM | Shanghai | Warren Wong | 13043 |----------|
| April 7, 2021 | 12:00 PM | EMEA | Andrew Smith | 13044 |----------|

# Event Checklist

## 10 Weeks Prior
- [ ] Reactor Content Team (RCT) opens a new issue for an upcoming local workshop filling in the Event Information and Workshop details tables. RCT member who owns content (content owner) this week will be the assignee.
- [ ] RCT and content owner discuss details of the event in the comments of this issue. RCT adds a _confirmed-event_ label and _local-digital-event_ tag to the issue and adds event to the __Local Digital Events__ project. Issue will appear in the __Upcoming Streams__ column
- [ ] RCT adds a _ready-for-marketing-edit_ label so the RCT can review title and description

## 8 Weeks Prior
- [ ] RCT does a review of title and description and removes the _ready-for-marketing-edit_ label 
- [ ] RCT moves the issue to the __Ready for SharePoint__ column
- [ ] RCT informs the Reactor Marketing Team of the new event information
- [ ] RCT confirms the speaker
- [ ] RCT provides speaker with assets (e.g. slides, overlays) and the survey data for the event
- [ ] RCT moves the issue to the __Content Preparation In Progress__ column

## 5 weeks Prior
- [ ] Content owner opens a PR with the new content for the workshop on the [customer-facing Reactor repository](https://github.com/microsoft/Reactors/tree/main/online-event-resources) and adds the _content-development_ and _local-digital-event_ tag to the PR. 
- [ ] RCT reviews, approves, and merges the changes
- [ ] RCT shares content with the vendor/instructor
- [ ] RCT provides vendor/instructor with assets (e.g. slides, overlays) and the survey data for the event
- [ ] RCT ensures that instructors have received an invite to the workshops and that PM team works with instructors to get setup in the platform (e.g. Teams, Teams Live Event)

## 1 Week Prior
- [ ] Connect vendor/instructor with the PM(s) for the locations where this event has been marketed
- [ ] Check-in with vendor/instructor that they are prepared for the workshop
- [ ] RCT moves the issue to the __Ready to Stream__ column

## During and After Stream
- [ ] Instructor (and optional moderator) run the stream
- [ ] _(Optional)_RCT submits recording to be uploaded to the Reactor YouTube and to be made available on LearnTV
- [ ] RCT moves issue to the __Stream Complete__ column

**Comment by mebaumb** [comment]
@meaghanlewis - What do you think this as a draft? 

Title: Build & Deploy Applications to Azure by Using GitHub Actions

Description: GitHub Actions are powerful workflow automations that support your projects success. You can use these automations in every step of the development project, from delivery to deployment. In this workshop we will learn how to use GitHub Actions to ensure continuous delivery (CD) of an application. We'll start by exploring options for triggering a CD GitHub Workflow, then control how the workflow is executed, and deploy the application to Microsoft Azure using a GitHub Deploy action. Afterwards, we'll do some data management while exploring GitHub Secrets and some additional GitHub Actions. 

To be successful in this workshop, you should have an understanding of GitHub Actions and continuous integration.

**Comment by meaghanlewis** [comment]
@mebaumb I made a few slight edits inline. Looks great to me, thanks for getting the jump on this.

**Comment by mebaumb** [comment]
@meaghanlewis - I will get this published, is this going to be beginner or intermediate content? 

Thanks!

**Comment by meaghanlewis** [comment]
@mebaumb this is intermediate content. thanks for publishing!

**Comment by mebaumb** [comment]
Waiting on short links and speakers, will publish tomorrow!

**Comment by mebaumb** [comment]
This has been published!

**Comment by meaghanlewis** [comment]
## Instructor notes

**Share this link with attendees at the beginning of the workshop: https://aka.ms/GithubActionsCD-8**

Let them know this is the content that will be covered, and that following the workshops they should step through to complete the module, including answering the knowledge check questions and rating the module.

This workshop covers using GitHub Actions to deploy to Azure. Be sure to have both a GitHub acount and Microsoft Azure account.

**Workshop steps:**

1. It would be helpful to start with an overview of Actions. Some helpful links include:
    - https://docs.github.com/actions
    - https://github.com/actions
    - https://github.com/features/actions
 
2. Highlight the options to trigger a CD workflow: https://docs.microsoft.com/learn/modules/github-actions-cd/2-deploy-azure

3. Walk through the Learning Lab for _GitHub Actions: Continuous Delivery with Azure together_: https://lab.github.com/githubtraining/github-actions:-continuous-delivery-with-azure

**Wrapping up:**

- At the end of the workshop, remind attendees to complete the knowledge check and summary so they can get a badge for completing the module.
- Also please encourage them to rate the module 😄
- Share out the survey link: https://aka.ms/Reactor/Survey with the event code (the Reactor contact on the call can give you this)

**Comment by AndrewSouthpaw** [comment]
When I ran through the Learning Lab on GitHub, I ran into issues with deploying to staging/production. The Docker CLI would complain about auth issues:

```
2021_04_07_lw1sdlwk00001O_docker.log:
2021-04-07T13:23:18.792Z INFO  - Pulling image: docker.pkg.github.com/andrewsouthpaw-demo/github-actions-continuous-delivery-azure/andrewsouthpaw-demo-azure-ttt:7072be859c6ed16715d21770e2937dbaf90151dc
2021-04-07T13:23:19.131Z ERROR - DockerApiException: Docker API responded with status code=InternalServerError, response={"message":"unauthorized: Your request could not be authenticated by the GitHub Packages service. Please ensure your access token is valid and has the appropriate scopes configured."}

2021-04-07T13:23:19.131Z ERROR - Pulling docker image docker.pkg.github.com/andrewsouthpaw-demo/github-actions-continuous-delivery-azure/andrewsouthpaw-demo-azure-ttt:7072be859c6ed16715d21770e2937dbaf90151dc failed:
2021-04-07T13:23:19.131Z INFO  - Pulling image from Docker hub: docker.pkg.github.com/andrewsouthpaw-demo/github-actions-continuous-delivery-azure/andrewsouthpaw-demo-azure-ttt:7072be859c6ed16715d21770e2937dbaf90151dc
2021-04-07T13:23:19.232Z ERROR - DockerApiException: Docker API responded with status code=InternalServerError, response={"message":"Get https://docker.pkg.github.com/v2/andrewsouthpaw-demo/github-actions-continuous-delivery-azure/andrewsouthpaw-demo-azure-ttt/manifests/7072be859c6ed16715d21770e2937dbaf90151dc: no basic auth credentials"}

2021-04-07T13:23:19.240Z ERROR - Image pull failed: Verify docker image configuration and credentials (if using private repository)
2021-04-07T13:23:19.328Z INFO  - Stopping site andrewsouthpaw-demo-ttt-app because it failed during startup.
```

The upshot is that the container wouldn't start, and if you visited the webapp it'd give an application error:

![image](https://user-images.githubusercontent.com/5187404/113874185-60b90d80-976a-11eb-86b2-3d0af5503c90.png)

I originally thought it was an issue with the `docker/login` action, which uses `secrets.GITHUB_TOKEN` as the auth token, because that was a problem for people in the past. However, supposedly that bug was fixed, and also using a fresh Personal Access Token didn't work either. 

I also confirmed the container settings on Azure had a valid token.

![image](https://user-images.githubusercontent.com/5187404/113873893-16d02780-976a-11eb-9667-1d6126c4a5be.png)

I'm not sure where would be the best place to report this issue, so I thought I'd raise it here first. I'm curious if other speakers had the same issue?

You can see a repo in progress [here](https://github.com/andrewsouthpaw-demo/github-actions-continuous-delivery-azure).

**Comment by meaghanlewis** [comment]
@AndrewSouthpaw thanks for reporting this issue! I'm not immediately sure where to raise this issue either. I think you had referenced this issue previously: https://github.com/githubtraining/continuous-delivery-azure/pull/34. I wonder if it's worth opening an issue in that repository with the error messages that you shared here, or adding a comment to this issue: https://github.com/githubtraining/continuous-delivery-azure/issues/29
